docs/data/data-grid/state/SaveAndRestoreStateInitialState.js

```diff
@@ -0,0 +1,70 @@
+import * as React from 'react';
+import { useDemoData } from '@mui/x-data-grid-generator';
+import { DataGridPremium, useGridApiRef } from '@mui/x-data-grid-premium';
+
+import { CircularProgress } from '@mui/material';
+import Box from '@mui/material/Box';
+
+function saveDataGridStateToLocalStorage(stateSnapshot) {
+  console.info('State update: Update the state snap in localStorage', stateSnapshot);
+  localStorage.setItem('dataGridState', JSON.stringify(stateSnapshot));
+}
+
+function getInitialStateFromLocalStorage() {
+  const stateFromLocalStorage = localStorage.getItem('dataGridState');
+  return new Promise((resolve) => {
+    if (!stateFromLocalStorage) {
+      resolve({});
+      return;
+    }
+    const state = JSON.parse(stateFromLocalStorage);
+    resolve(state);
+  });
+}
+
+export default function SaveAndRestoreStateInitialState() {
+  const apiRef = useGridApiRef();
+  const { data, loading } = useDemoData({
+    dataSet: 'Commodity',
+    rowLength: 100,
+    maxColumns: 10,
+  });
+
+  const [initialState, setInitialState] = React.useState();
+
+  React.useEffect(() => {
+    const getSnapshot = async () => {
+      const snapshotFromLocalStorage = await getInitialStateFromLocalStorage();
```

Comment:
I feel like useEffect and async make it a bit more complex than it needs to be 🙂
What do you think about using the initializer function for the initial state?

```tsx
const [initialState, setInitialState] = React.useState(() => {
  const stateFromLocalStorage = localStorage.getItem('dataGridState');
  return stateFromLocalStorage || {};
});
```

Also, adding error handling in case localStorage is not available would be great!

Reply:
I did simplify the example code a bit.

One thing that is not possible is the extraction into the state initializer, since we need to wait for `localStorage` to be initialized (limitation of nextjs)

Reply:
Ah right, I think I already had the same discussion about state initializer and local storage 😅